flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/writer/FileWriterBucket.java

```diff
@@ -0,0 +1,329 @@
+
+	private long partCounter;
+
+	@Nullable
+	private InProgressFileRecoverable inProgressFileToCleanup;
+
+	@Nullable
+	private InProgressFileWriter<IN, BucketID> inProgressPart;
+
+	/**
+	 * Constructor to create a new empty bucket.
+	 */
+	private FileWriterBucket(
+			BucketID bucketId,
+			Path bucketPath,
+			String uniqueId,
+			BucketWriter<IN, BucketID> bucketWriter,
+			RollingPolicy<IN, BucketID> rollingPolicy,
+			OutputFileConfig outputFileConfig) {
+		this.bucketId = checkNotNull(bucketId);
+		this.bucketPath = checkNotNull(bucketPath);
+		this.uniqueId = checkNotNull(uniqueId);
+		this.bucketWriter = checkNotNull(bucketWriter);
+		this.rollingPolicy = checkNotNull(rollingPolicy);
+		this.outputFileConfig = checkNotNull(outputFileConfig);
+
+		this.partCounter = 0;
+	}
+
+	/**
+	 * Constructor to restore a bucket from checkpointed state.
+	 */
+	private FileWriterBucket(
+			String uniqueId,
+			BucketWriter<IN, BucketID> partFileFactory,
+			RollingPolicy<IN, BucketID> rollingPolicy,
+			FileWriterBucketState<BucketID> bucketState,
+			OutputFileConfig outputFileConfig) throws IOException {
+
+		this(
+				bucketState.getBucketId(),
+				bucketState.getBucketPath(),
+				uniqueId,
+				partFileFactory,
+				rollingPolicy,
+				outputFileConfig);
+
+		restoreInProgressFile(bucketState);
+	}
+
+	private void restoreInProgressFile(FileWriterBucketState<BucketID> state) throws IOException {
+		if (!state.hasInProgressFileRecoverable()) {
+			return;
+		}
+
+		// we try to resume the previous in-progress file
+		InProgressFileWriter.InProgressFileRecoverable inProgressFileRecoverable =
+				state.getInProgressFileRecoverable();
+
+		if (bucketWriter.getProperties().supportsResume()) {
+			inProgressPart = bucketWriter.resumeInProgressFileFrom(
+					bucketId, inProgressFileRecoverable, state.getInProgressFileCreationTime());
+		} else {
+			pendingFiles.add(inProgressFileRecoverable);
+		}
+	}
+
+	public BucketID getBucketId() {
+		return bucketId;
+	}
+
+	public Path getBucketPath() {
+		return bucketPath;
+	}
+
+	public long getPartCounter() {
+		return partCounter;
+	}
+
+	public boolean isActive() {
+		return inProgressPart != null || inProgressFileToCleanup != null || pendingFiles.size() > 0;
+	}
+
+	void merge(final FileWriterBucket<IN, BucketID> bucket) throws IOException {
+		checkNotNull(bucket);
+		checkState(Objects.equals(bucket.bucketPath, bucketPath));
+
+		bucket.closePartFile();
+		pendingFiles.addAll(bucket.pendingFiles);
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Merging buckets for bucket id={}", bucketId);
+		}
+	}
+
+	void write(IN element) throws IOException {
+		long now = System.currentTimeMillis();
+		if (inProgressPart == null || rollingPolicy.shouldRollOnEvent(inProgressPart, element)) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug(
+						"closing in-progress part file for bucket id={} due to element {}.",
```

Comment:
I think that here it should be "Opening new in-progress part file for..." because if there is no part file, then we are not closing anything.